File: perl/gc.pl
#!/usr/bin/perl
use strict;
use warnings;

open(my $fh, '<', "chry_multiplied.fa") or die $!;
my $gc = 0;
my $at = 0;
while (my $line = <$fh>) {
    if ($line  =~ /^>/) {
        next;
    }

    $gc += ($line =~ tr/GC//);
    $at += ($line =~ tr/AT//);
}

my $gc_ration = $gc / ($gc + $at);
print "GC ration (gc/gc+at)): $gc_ration\n";

File: node/gc.js
#!/usr/bin/env node
const fs = require('fs')
const readline = require('readline')

const rl = readline.createInterface({
  input: fs.createReadStream('chry_multiplied.fa')
})

// https://stackoverflow.com/a/60361745 - performance recommendation
function countCharacter (str, char) {
  var start = 0
  var count = 0
  while ((start = str.indexOf(char, start) + 1) !== 0) {
    count++
  }
  return count
}

var gc = 0
var total = 0

rl.on('line', (line) => {
  if (!(line.startsWith('>'))) {
    var countA = countCharacter(line, 'A')
    var countC = countCharacter(line, 'C')
    var countG = countCharacter(line, 'G')
    var countT = countCharacter(line, 'T')
    gc += countC + countG
    total += countA + countC + countG + countT
  }
})

rl.on('close', () => {
  var gcFraction = gc / total
  console.log(gcFraction * 100)
})
